### PR TITLE
Add support to validate_all for STACObject

### PR DESCRIPTION
**Pull request by chuckwondo**
**Related Issue(s):**

- #1206 

**Description:**

- add `validate_all_dict` function for validating objects as dicts (i.e., rename original `validate_all` function)
- deprecate support for passing dict objects to `validate_all` (use new `validate_all_dict` function instead)
- add support for passing `STACObject` to `validate_all` (and prohibit passing a value for `href` when passing a `STACObject`)

**PR Checklist:**

- [x] `pre-commit` hooks pass locally
- [x] Tests pass (run `scripts/test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] This PR maintains or improves overall codebase code coverage.
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/main/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

File: tests/validation/test_validate.py [hunk 1]
```diff
@@ -80,14 +79,46 @@ def test_validate_error_contains_href(self) -> None:
                 assert get_opt(item.get_self_href()) in str(e)
                 raise e
 
+    @pytest.mark.vcr()
+    def test_validate_all_deprecated_dict_arg(self) -> None:
+        catalog = TestCases.case_1()
+        catalog.fully_resolve()
```
Comment (gadomski) [inline tests/validation/test_validate.py:85]: Did you check to see if these are still required?

**Review by gadomski — CHANGES_REQUESTED** [review]

File: tests/validation/test_validate.py (hunk 1 above)
Comment (chuckwondo) [reply tests/validation/test_validate.py]: Ah, yes, I was able to remove the call to `fully_resolve`.

File: tests/validation/test_validate.py [hunk 2]
```diff
@@ -80,14 +79,45 @@ def test_validate_error_contains_href(self) -> None:
...
+    def test_validate_all_deprecated_dict_arg(self) -> None:
+        catalog = TestCases.case_1()
+
+        with pytest.warns(DeprecationWarning, match="use validate_all_dict"):
+            pystac.validation.validate_all(catalog.to_dict(), catalog.get_self_href())
+
+    @pytest.mark.vcr()
+    def test_validate_all_deprecated_dict_arg_missing_href(self) -> None:
+        catalog = TestCases.case_1()
+        catalog.fully_resolve()
```
Comment (gadomski) [inline tests/validation/test_validate.py:92]:
```suggestion
```

File: tests/validation/test_validate.py [hunk 3]
```diff
@@ -80,14 +79,45 @@ def test_validate_error_contains_href(self) -> None:
...
+
+        with pytest.warns(DeprecationWarning, match="use validate_all_dict"):
+            with pytest.raises(ValueError, match="href must be set"):
+                pystac.validation.validate_all(catalog.to_dict())
+
+    @pytest.mark.vcr()
+    def test_validate_all_unexpected_href(self) -> None:
+        catalog = TestCases.case_1()
+        catalog.fully_resolve()
```
Comment (gadomski) [inline tests/validation/test_validate.py:101]:
```suggestion
```

File: tests/validation/test_validate.py [hunk 4]
```diff
@@ -80,14 +79,45 @@ def test_validate_error_contains_href(self) -> None:
...
+
+        with pytest.raises(ValueError, match="href must be None"):
+            pystac.validation.validate_all(catalog, catalog.get_self_href())
+
+    @pytest.mark.vcr()
+    def test_validate_all(self) -> None:
+        catalog = TestCases.case_1()
+        catalog.fully_resolve()
```
Comment (gadomski) [inline tests/validation/test_validate.py:109]:
```suggestion
```

**Review by gadomski — APPROVED** [review]

**Pull request merged**